python-spec/src/somacore/query/fast_csr.py

```diff
@@ -0,0 +1,291 @@
+            row_rng_val,
+        )
+    return None
+
+
+@numba.jit(nopython=True, nogil=True)  # type: ignore
+def _finalize_indptr(indptr):
+    prev = 0
+    for r in range(len(indptr)):
+        t = indptr[r]
+        indptr[r] = prev
+        prev = t
+    return None
+
+
+def _select_dtype(
+    maxval: int,
+) -> Type[np.signedinteger]:
+    """
+    Ascertain the "best" dtype for a zero-based index. Given our
+    goal of minimizing memory use, "best" is currently defined as
+    smallest.
+    """
+    if maxval > np.iinfo(np.int32).max:
+        return np.int64
+    else:
+        return np.int32
+
+
+def _reindex_and_cast(
+    index: pd.Index, ids: npt.NDArray[np.int64], target_dtype: npt.DTypeLike
+) -> npt.NDArray[np.int64]:
+    return cast(
+        npt.NDArray[np.int64], index.get_indexer(ids).astype(target_dtype, copy=False)
+    )
+
+
+class CSRAccumulator:
+    """
+    Fast accumulator of a CSR, based upon COO input.
+    """
+
+    def __init__(
+        self,
+        obs_joinids: npt.NDArray[np.int64],
+        var_joinids: npt.NDArray[np.int64],
+        pool: futures.Executor,
+    ):
+        self.obs_joinids = obs_joinids
+        self.var_joinids = var_joinids
+        self.pool = pool
+
+        self.shape: Tuple[int, int] = (len(self.obs_joinids), len(self.var_joinids))
+        self.obs_indexer: pd.Index = pd.Index(self.obs_joinids)
+        self.var_indexer: pd.Index = pd.Index(self.var_joinids)
+        self.row_length: npt.NDArray[np.int64] = np.zeros(
+            (self.shape[0],), dtype=_select_dtype(self.shape[1])
+        )
+
+        # COO accumulated chunks, stored as list of triples (data, row_ind, col_ind)
+        self.coo_chunks: List[
+            Tuple[
+                npt.NDArray[np.number],
+                npt.NDArray[np.integer],
+                npt.NDArray[np.integer],
+            ]
+        ] = []
+
+    def append(
+        self, data: pa.Array, row_joinids: pa.Array, col_joinids: pa.Array
+    ) -> None:
+        """
+        At accumulation time, do several things:
+        * re-index to positional indices, and if possible, cast to smaller dtype
+          to minimize memory footprint (at cost of some amount of time)
+        * accumulate column counts by row, i.e., build the basis of the indptr
+        * cache the tuple of data, row, col
+        """
+        rows_future = self.pool.submit(
+            _reindex_and_cast,
+            self.obs_indexer,
+            row_joinids.to_numpy(),
+            _select_dtype(self.shape[0]),
+        )
+        cols_future = self.pool.submit(
+            _reindex_and_cast,
+            self.var_indexer,
+            col_joinids.to_numpy(),
+            _select_dtype(self.shape[1]),
+        )
+        row_ind = rows_future.result()
+        col_ind = cols_future.result()
+        self.coo_chunks.append((data.to_numpy(), row_ind, col_ind))
+        _accum_row_length(self.row_length, row_ind)
+
+    def finalize(
+        self,
+    ) -> Tuple[
```

Comment:
Would it make sense to return this as an attrs class or namedtuple so that these fields can have a name rather than / in addition to being positional?

Reply:
the overhead didn't seem worth it given that it is a private interface.  But sure, why not ... :-)

Reply:
I used a namedtuple.  LMK if this was along the lines of what you were thinking.